lib/cnet/chnl/chnl_open.c

```diff
@@ -0,0 +1,199 @@
+ * to have more then one string type with different number of fields.
+ */
+static chnl_open_t open_types[] = {
+    {"udp-listen",   UDP4_LISTEN,  2}, /* udp-listen:<port> */
+    {"udp-listen",   UDP4_LISTEN,  3}, /* udp-listen:<ipaddr>:<port> */
+    {"udp-connect",  UDP4_CONNECT, 3}, /* udp->connect:<ipaddr>:<port> */
+
+    {"udp4-listen",  UDP4_LISTEN,  2}, /* udp4-listen:<port> */
+    {"udp4-listen",  UDP4_LISTEN,  3}, /* udp4-listen:<ipaddr>:<port> */
+    {"udp4-connect", UDP4_CONNECT, 3}, /* udp4->connect:<ipaddr>:<port> */
+
+    {"udp6-listen",  UDP6_LISTEN,  2}, /* udp6-listen:<port> */
+    {"udp6-listen",  UDP6_LISTEN,  3}, /* udp6-listen:<ipaddr>:<port> */
+    {"udp6-connect", UDP6_CONNECT, 3}, /* udp6->connect:<ipaddr>:<port> */
+
+    {"tcp-listen",   TCP4_LISTEN,  2}, /* tcp-listen:<port> */
+    {"tcp-listen",   TCP4_LISTEN,  3}, /* tcp-listen:<ipaddr>:<port> */
+    {"tcp-connect",  TCP4_CONNECT, 3}, /* tcp-connect:<ipaddr>:<port> */
+
+    {"tcp4-listen",  TCP4_LISTEN,  2}, /* tcp4-listen:<port> */
+    {"tcp4-listen",  TCP4_LISTEN,  3}, /* tcp4-listen:<ipaddr>:<port> */
+    {"tcp4-connect", TCP4_CONNECT, 3}, /* tcp4-connect:<ipaddr>:<port> */
+
+    {"tcp6-listen",  TCP6_LISTEN,  2}, /* tcp6-listen:<port> */
+    {"tcp6-listen",  TCP6_LISTEN,  3}, /* tcp6-listen:<ipaddr>:<port> */
+    {"tcp6-connect", TCP6_CONNECT, 3}, /* tcp6-connect:<ipaddr>:<port> */
+    {NULL, MAX_OPEN_TYPES, 0}
+    };
+// clang-format on
+
+static int
+create_chnl(int otype, int domain, int type, const char *name, int port, int flags, chnl_cb_t cb)
+{
+    int cd;
+    uint32_t opt;
+
+    if (!cb)
+        CNE_ERR_RET("channel callback is NULL\n");
+
+    cd = channel(domain, type, 0, cb);
+    if (cd < 0)
+        CNE_ERR_RET("channel call failed\n");
+
+    opt = 1;
+    chnl_set_opt(cd, SO_CHANNEL, SO_REUSEADDR, &opt, sizeof(uint32_t));
+
+    opt = (flags & CHNL_ENABLE_UDP_CHECKSUM) ? 1 : 0;
+    chnl_set_opt(cd, 0, SO_UDP_CHKSUM, &opt, sizeof(uint32_t));
+
+    if (name) {
+        if (domain == AF_INET) {
+            struct in_caddr addr;
+
+            in_caddr_zero(&addr);
+
+            if (inet_pton(AF_INET, name, (void *)&addr.cin_addr.s_addr) != 1)
+                CNE_ERR_RET("Unable to convert IP4 address to network order\n");
+            addr.cin_family = domain;
+            addr.cin_len    = sizeof(struct in_addr);
+            addr.cin_port   = htobe16(port);
+
+            if (otype == TCP4_LISTEN || otype == UDP4_LISTEN) {
+                if (chnl_bind(cd, (struct sockaddr *)&addr, sizeof(struct in_caddr)) == -1)
+                    CNE_ERR_RET("chnl_bind() failed\n");
+            } else if (otype == TCP4_CONNECT || otype == UDP4_CONNECT) {
+                if (chnl_connect(cd, (struct sockaddr *)&addr, sizeof(struct in_caddr)))
+                    CNE_ERR_RET("chnl_connect() failed\n");
+            }
+        } else if (domain == AF_INET6)
+            CNE_ERR_RET("IPv6 is not supported\n");
```

Comment:
I see you don't need to validate "domain" becuase the previous channel() call will fail if there is no matching protosw entry.

Reply:
I would like to keep that check as some point when INET6 is valid and we need to continue some checks for this domain.

Reply:
Yeah, we need it.

I meant it's ok that we don't check if domain != AF_INET and domain != AF_INET6. That will already be handled in the call to channel(), which will fail if the applicatno sets domain to 12345 or something silly like that.